lib/common/pool.c

```diff
@@ -89,30 +91,32 @@ static void* POOL_thread(void* opaque) {
             job.function(job.opaque);
 
             /* If the intended queue size was 0, signal after finishing job */
+            ZSTD_pthread_mutex_lock(&ctx->queueMutex);
+            ctx->numThreadsBusy--;
             if (ctx->queueSize == 1) {
-                ZSTD_pthread_mutex_lock(&ctx->queueMutex);
-                ctx->numThreadsBusy--;
-                ZSTD_pthread_mutex_unlock(&ctx->queueMutex);
                 ZSTD_pthread_cond_signal(&ctx->queuePushCond);
```

Comment:
We always need to signal now, or at least signal if `numThreadsBusy == theadLimit` before decrementing.

Reply:
That comment isn't quite right. However, I think we should move the increment/decrement of `numThreadsBusy`.

Shouldn't we decrement right before we actually sleep (and start with it at one).

Reply:
Another strategy could be to have a local variable `isActive`
which knows if the thread was just active before or was awaken.
It would then be able to "fix" the real nb of active threads, or the decision to do `numThreads++`.

But this alternative design feels fragile, easier to break.